django/db/models/functions/math.py

```diff
@@ -0,0 +1,164 @@
+import math
+import sys
+
+from django.db.models import DecimalField, FloatField, Func, Transform
+from django.db.models.functions import Cast
+
+
+class DecimalInputMixin:
+
+    def as_postgresql(self, compiler, connection):
+        # Cast FloatField to DecimalField as PostgreSQL doesn't support the
+        # following function signatures by default:
+        # - LOG(double precision, double precision)
+        # - MOD(double precision, double precision)
+        output_field = DecimalField(decimal_places=sys.float_info.dig, max_digits=1000)
+        clone = self.copy()
+        clone.set_source_expressions([
+            Cast(expression, output_field) if isinstance(expression.output_field, FloatField)
+            else expression for expression in self.get_source_expressions()
+        ])
+        return clone.as_sql(compiler, connection)
+
+
+class OutputFieldMixin:
+
+    def _resolve_output_field(self):
+        sources = self.get_source_expressions()
+        if any(isinstance(s.output_field, DecimalField) for s in sources):
+            return DecimalField()
+        else:
+            return FloatField()
+
+
+class Abs(Transform):
+    function = 'ABS'
+    lookup_name = 'abs'
+
+
+class ACos(OutputFieldMixin, Transform):
+    function = 'ACOS'
+    lookup_name = 'acos'
+
+
+class ASin(OutputFieldMixin, Transform):
+    function = 'ASIN'
+    lookup_name = 'asin'
+
+
+class ATan(OutputFieldMixin, Transform):
+    function = 'ATAN'
+    lookup_name = 'atan'
+
+
+class ATan2(OutputFieldMixin, Func):
```

Comment:
Support was added for `ATan2()` in SpatiaLite 4.3.0, but it fails the tests.

```
======================================================================
FAIL: test_decimal (db_functions.math.test_atan2.ATan2Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/nick/Sources/django/tests/db_functions/math/test_atan2.py", line 16, in test_decimal
    self.assertAlmostEqual(obj.n_atan2, Decimal(math.atan2(obj.n1, obj.n2)))
  File "/usr/lib/python3.6/unittest/case.py", line 878, in assertAlmostEqual
    raise self.failureException(msg)
AssertionError: Decimal('2.70662570254421') != Decimal('-1.135829375749316749732997777755372226238250732421875') within 7 places

======================================================================
FAIL: test_float (db_functions.math.test_atan2.ATan2Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/nick/Sources/django/tests/db_functions/math/test_atan2.py", line 22, in test_float
    self.assertAlmostEqual(obj.f_atan2, math.atan2(obj.f1, obj.f2))
  File "/usr/lib/python3.6/unittest/case.py", line 878, in assertAlmostEqual
    raise self.failureException(msg)
AssertionError: 3.128393420165654 != -1.5575970933707572 within 7 places
```
Strangely, after inverting the order of the arguments - similar to the issue with `Log()` - the float test passes and the integer test passes when casting the inputs to floats.

So the following method needs to be added:
```python
    def as_sqlite(self, compiler, connection):
        if not getattr(connection.ops, 'spatialite', False):
            return self.as_sql(compiler, connection)
        if connection.ops.spatial_version < (4, 3, 0):
            return self.as_sql(compiler, connection)
        # As documented this function ought to be ATan2(y, x) returning the
        # inverse tangent of y / x, but it turns out that, unlike all other
        # backends, the function is actually ATan2(x, y). In addition, it
        # doesn't handle integer inputs properly, so these need to be casted to
        # floats.
        # Although the spatialite backend inherits from the sqlite3 backend,
        # the spatialite function takes precedence over the one created using
        # create_function().
        clone = self.copy()
        clone.set_source_expressions([
            Cast(expression, FloatField()) if not isinstance(expression.output_field, FloatField)
            else expression for expression in self.get_source_expressions()[::-1]
        ])
        return clone.as_sql(compiler, connection)
```

Reply:
Sorry - I should clarify - the tests were failing for float and decimal.
This was fixed by inverting the order of the arguments, but that caused integer to fail.
Casting the arguments to float seemed to then make the integer test pass.

If somebody else could scrutinise my suggested fix, that would be great.

Reply:
@pope1ni Thank you for the instructions and detailed analysis. It is weird that the Atan2 function cannot pass SpatiaLite tests. Should we consider to use as_sql(...) for all SpatiaLite Atan2 functions?

Reply:
I'm not sure why the change I suggested to add `ATan2.as_sqlite()` targeting SpatiaLite >= 4.3.0 resolved this issue for me locally and not on Jenkins.

Could you confirm the version of SpatiaLite on the xenial and bionic boxes, @timgraham?

(Maybe a build just needs to be re-triggered.)